File: src-tauri/src/notifications/webhook.rs
//! Webhook delivery ported from `internal/notifications/webhook.go`:
//! template expansion, the SSRF allowlist, retry-with-backoff, and the real
//! HTTP dispatch (via `ureq` 3).
//!
//! SSRF hardening reproduced from the Go fix:
//!   1. `parse_url` rejects any authority containing `@` (userinfo). Otherwise
//!      `https://hooks.slack.com:443@127.0.0.1/x` passes the host allowlist while
//!      the client connects to `127.0.0.1`.
//!   2. The HTTP client does NOT follow redirects (`max_redirects(0)`) — a 3xx
//!      to an internal address would be a redirect-SSRF hole. `check_ssrf`
//!      validates only the initial URL.

use std::net::IpAddr;
use std::sync::atomic::{AtomicU32, Ordering};
use std::time::Duration;

use serde::{Deserialize, Serialize};

const ALLOWED_DISCORD_PATH_PREFIXES: &[&str] = &["/api/webhooks/"];

const MAX_ATTEMPTS: u32 = 3;
const RETRY_DELAYS_MS: &[u64] = &[1000, 2000, 4000];

// ── Types ───────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct WebhookEndpoint {
    pub id: String,
    pub label: String,
    pub url: String,
    pub template: String,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WebhookContext {
    pub session_id: String,
    pub tool: String,
    pub cost: f64,
    pub summary: String,
}

#[derive(Debug)]
pub enum WebhookError {
    SsrfRejected(String),
    InvalidUrl(String),
    HttpStatus(u16),
    NetworkFailure(String),
}

impl std::fmt::Display for WebhookError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            WebhookError::SsrfRejected(why) => write!(f, "SSRF guard rejected: {why}"),
            WebhookError::InvalidUrl(why) => write!(f, "Invalid URL: {why}"),
            WebhookError::HttpStatus(code) => write!(f, "HTTP {code}"),
            WebhookError::NetworkFailure(why) => write!(f, "Network: {why}"),
        }
    }
}

impl std::error::Error for WebhookError {}

/// Result of an HTTP attempt — drives the retry loop's back-off decision.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AttemptOutcome {
    Success,
    Retryable,
    Permanent,
}

/// Abstracts the HTTP layer so tests can substitute a fake transport.
pub trait WebhookTransport: Send + Sync {
    fn send(&self, url: &str, body: &str) -> AttemptOutcome;
}

/// Counts attempts (mirrors `RetryStats`).
#[derive(Debug, Default)]
pub struct RetryStats {
    pub attempts: AtomicU32,
}

// ── SSRF allowlist ────────────────────────────────────────────────────────────

fn host_allowed(host: &str, path: &str) -> bool {
    let host_lc = host.to_ascii_lowercase();
    if host_lc == "hooks.slack.com" {
        return true;
    }
    if host_lc == "discord.com" || host_lc == "discordapp.com" {
        return ALLOWED_DISCORD_PATH_PREFIXES
            .iter()
            .any(|prefix| path.starts_with(prefix));
    }
    false
}

fn ip_is_private(ip: IpAddr) -> bool {
    match ip {
        IpAddr::V4(v4) => {
            let o = v4.octets();
            v4.is_loopback()
                || v4.is_private()
                || v4.is_link_local() // 169.254.0.0/16 (link-local unicast)
                || (o[0] == 224 && o[1] == 0 && o[2] == 0) // link-local multicast
                || o[0] == 0
                || (o[0] == 169 && o[1] == 254)
        }
        // IPv6: loopback, unspecified, or ULA (fc00::/7).
        IpAddr::V6(v6) => {
            v6.is_loopback() || v6.is_unspecified() || (v6.octets()[0] & 0xfe) == 0xfc
        }
    }
}

#[derive(Debug)]
struct ParsedUrl<'a> {
    scheme: &'a str,
    host: &'a str,
    path: &'a str,
}

fn parse_url(url: &str) -> Result<ParsedUrl<'_>, WebhookError> {
    let (scheme, rest) = url
        .split_once("://")
        .ok_or_else(|| WebhookError::InvalidUrl("missing scheme".to_string()))?;
    if scheme != "https" && scheme != "http" {
        return Err(WebhookError::InvalidUrl(format!("scheme {scheme} not allowed")));
    }
    let (host_with_port, path) = match rest.find('/') {
        Some(idx) => (&rest[..idx], &rest[idx..]),
        None => (rest, "/"),
    };
    // Reject userinfo: "user@host" would let the allowlist see an approved host
    // while the client connects to the part after '@' (SSRF bypass). No
    // legitimate webhook URL carries userinfo.
    if host_with_port.contains('@') {
        return Err(WebhookError::InvalidUrl(
            "userinfo not allowed in webhook URL".to_string(),
        ));
    }
    // Strip port. Only when there are no brackets (bracket ⇒ IPv6 literal).
    let host = if host_with_port.contains('[') {
        host_with_port
    } else {
        match host_with_port.rfind(':') {
            Some(i) => &host_with_port[..i],
            None => host_with_port,
        }
    };
    Ok(ParsedUrl { scheme, host, path })
}

/// Validates a URL against the SSRF allowlist (mirrors `webhook.go:CheckSSRF`).
pub fn check_ssrf(url: &str) -> Result<(), WebhookError> {
    let parsed = parse_url(url)?;
    if parsed.scheme == "http" {
        return Err(WebhookError::SsrfRejected("http scheme blocked".to_string()));
    }
    if !host_allowed(parsed.host, parsed.path) {
        return Err(WebhookError::SsrfRejected(format!(
            "host {} not on allowlist",
            parsed.host
        )));
    }
    // Reject literal private/loopback/link-local/ULA IP hosts.
    if let Ok(ip) = parsed.host.parse::<IpAddr>() {
        if ip_is_private(ip) {
            return Err(WebhookError::SsrfRejected("literal IP host".to_string()));
        }
    }
    Ok(())
}

// ── Template expansion ────────────────────────────────────────────────────────

pub fn expand_template(template: &str, ctx: &WebhookContext) -> String {
    template
        .replace("{session_id}", &ctx.session_id)
        .replace("{tool}", &ctx.tool)
        .replace("{cost}", &format!("{:.4}", ctx.cost))
        .replace("{summary}", &ctx.summary)
}

// ── Retry loop ────────────────────────────────────────────────────────────────

/// Runs the transport with backoff. `sleeper` is injectable for tests.
pub fn dispatch_with_retry(
    transport: &dyn WebhookTransport,
    url: &str,
    body: &str,
    sleeper: &dyn Fn(Duration),
    stats: &RetryStats,
) -> Result<(), WebhookError> {
    for attempt in 0..MAX_ATTEMPTS {
        stats.attempts.fetch_add(1, Ordering::SeqCst);
        match transport.send(url, body) {
            AttemptOutcome::Success => return Ok(()),
            AttemptOutcome::Permanent => return Err(WebhookError::HttpStatus(0)),
            AttemptOutcome::Retryable => {
                if attempt + 1 >= MAX_ATTEMPTS {
                    return Err(WebhookError::NetworkFailure(
                        "Exceeded max retry attempts".to_string(),
                    ));
                }
                let delay = RETRY_DELAYS_MS.get(attempt as usize).copied().unwrap_or(4000);
                sleeper(Duration::from_millis(delay));
            }
        }
    }
    Err(WebhookError::NetworkFailure("retry exhausted".to_string()))
}

/// Validates, expands the template, then dispatches (mirrors `DispatchWebhook`).
pub fn dispatch_webhook(
    transport: &dyn WebhookTransport,
    endpoint: &WebhookEndpoint,
    ctx: &WebhookContext,
) -> Result<(), WebhookError> {
    check_ssrf(&endpoint.url)?;
    let body = expand_template(&endpoint.template, ctx);
    let stats = RetryStats::default();
    dispatch_with_retry(
        transport,
        &endpoint.url,
        &body,
        &|d| std::thread::sleep(d),
        &stats,
    )
}

// ── Default (real) HTTP transport ─────────────────────────────────────────────

/// Sends a real POST using `ureq` 3. Redirects are NOT followed: `check_ssrf`
/// validates only the initial URL, so following a 3xx to an internal address
/// would be a redirect-SSRF hole.
pub struct HttpTransport {
    agent: ureq::Agent,
}

impl HttpTransport {
    /// Creates a transport with a 10-second timeout and redirects disabled
    /// (mirrors `webhook.go:NewHTTPTransport`).
    pub fn new() -> Self {
        let agent: ureq::Agent = ureq::Agent::config_builder()
            .timeout_global(Some(Duration::from_secs(10)))
            .max_redirects(0)
            .http_status_as_error(false)
            .build()
            .into();
        Self { agent }
    }
}

impl Default for HttpTransport {
    fn default() -> Self {
        Self::new()
    }
}

impl WebhookTransport for HttpTransport {
    fn send(&self, url: &str, body: &str) -> AttemptOutcome {
        match self
            .agent
            .post(url)
            .header("Content-Type", "application/json")
            .header("User-Agent", "claude-devtools/webhook")
            .send(body)
        {
            Ok(resp) => {
                let code = resp.status().as_u16();
                if (200..300).contains(&code) {
                    AttemptOutcome::Success
                } else if code >= 500 {
                    AttemptOutcome::Retryable
                } else {
                    AttemptOutcome::Permanent
                }
            }
            Err(_) => AttemptOutcome::Retryable,
        }
    }
}

#[cfg(test)]
#[path = "webhook_tests.rs"]
mod webhook_tests;

File: internal/notifications/webhook.go
// webhook.go.
// Template expansion, SSRF allowlist, retry-with-backoff, HTTP dispatch.
// Uses stdlib net/http for the actual POST — no reqwest equivalent needed.
package notifications

import (
	"bytes"
	"fmt"
	"net"
	"net/http"
	"strings"
	"sync/atomic"
	"time"
)

// ── Types ─────────────────────────────────────────────────────────────────────

// WebhookEndpoint mirrors webhook.rs WebhookEndpoint.
type WebhookEndpoint struct {
	ID       string `json:"id"`
	Label    string `json:"label"`
	URL      string `json:"url"`
	Template string `json:"template"`
}

// WebhookContext mirrors webhook.rs WebhookContext.
type WebhookContext struct {
	SessionID string  `json:"sessionId"`
	Tool      string  `json:"tool"`
	Cost      float64 `json:"cost"`
	Summary   string  `json:"summary"`
}

// WebhookError mirrors webhook.rs WebhookError.
type WebhookError struct {
	Kind string
	Why  string
	Code int
}

func (e *WebhookError) Error() string {
	switch e.Kind {
	case "ssrf":
		return fmt.Sprintf("SSRF guard rejected: %s", e.Why)
	case "invalidUrl":
		return fmt.Sprintf("Invalid URL: %s", e.Why)
	case "httpStatus":
		return fmt.Sprintf("HTTP %d", e.Code)
	case "network":
		return fmt.Sprintf("Network: %s", e.Why)
	}
	return e.Why
}

func ssrfError(why string) *WebhookError    { return &WebhookError{Kind: "ssrf", Why: why} }
func urlError(why string) *WebhookError     { return &WebhookError{Kind: "invalidUrl", Why: why} }
func statusError(code int) *WebhookError    { return &WebhookError{Kind: "httpStatus", Code: code} }
func networkError(why string) *WebhookError { return &WebhookError{Kind: "network", Why: why} }

// AttemptOutcome mirrors webhook.rs AttemptOutcome.
type AttemptOutcome int

const (
	AttemptSuccess   AttemptOutcome = iota
	AttemptRetryable AttemptOutcome = iota
	AttemptPermanent AttemptOutcome = iota
)

// WebhookTransport abstracts the HTTP layer for testing.
// Mirrors webhook.rs WebhookTransport.
type WebhookTransport interface {
	Send(url, body string) AttemptOutcome
}

// RetryStats counts attempts (mirrors webhook.rs RetryStats).
type RetryStats struct {
	Attempts atomic.Int32
}

// ── SSRF allowlist ────────────────────────────────────────────────────────────

var allowedDiscordPathPrefixes = []string{"/api/webhooks/"}

func hostAllowed(host, path string) bool {
	h := strings.ToLower(host)
	if h == "hooks.slack.com" {
		return true
	}
	if h == "discord.com" || h == "discordapp.com" {
		for _, prefix := range allowedDiscordPathPrefixes {
			if strings.HasPrefix(path, prefix) {
				return true
			}
		}
		return false
	}
	return false
}

func ipIsPrivate(ip net.IP) bool {
	if ip4 := ip.To4(); ip4 != nil {
		return ip.IsLoopback() || ip.IsPrivate() || ip.IsLinkLocalUnicast() ||
			ip.IsLinkLocalMulticast() || ip4[0] == 0 ||
			(ip4[0] == 169 && ip4[1] == 254)
	}
	// IPv6: loopback, unspecified, or ULA (fc00::/7)
	return ip.IsLoopback() || ip.IsUnspecified() || (ip[0]&0xfe) == 0xfc
}

// parsedURL holds minimal URL decomposition.
type parsedURL struct {
	scheme string
	host   string
	path   string
}

func parseURL(rawURL string) (*parsedURL, *WebhookError) {
	idx := strings.Index(rawURL, "://")
	if idx < 0 {
		return nil, urlError("missing scheme")
	}
	scheme := rawURL[:idx]
	if scheme != "https" && scheme != "http" {
		return nil, urlError(fmt.Sprintf("scheme %s not allowed", scheme))
	}
	rest := rawURL[idx+3:]
	// Split host from path.
	pathIdx := strings.Index(rest, "/")
	var hostWithPort, path string
	if pathIdx >= 0 {
		hostWithPort = rest[:pathIdx]
		path = rest[pathIdx:]
	} else {
		hostWithPort = rest
		path = "/"
	}
	// Reject userinfo: "user@host" would let the allowlist see an approved host
	// while net/http connects to the part after '@' (SSRF bypass). No legitimate
	// webhook URL carries userinfo.
	if strings.ContainsRune(hostWithPort, '@') {
		return nil, urlError("userinfo not allowed in webhook URL")
	}
	// Strip port.
	host := hostWithPort
	if i := strings.LastIndex(hostWithPort, ":"); i >= 0 {
		// Only strip if what follows looks like a port number (no brackets = not IPv6).
		if !strings.Contains(hostWithPort, "[") {
			host = hostWithPort[:i]
		}
	}
	return &parsedURL{scheme: scheme, host: host, path: path}, nil
}

// CheckSSRF validates a URL against the SSRF allowlist.
// Mirrors webhook.rs::check_ssrf.
func CheckSSRF(url string) *WebhookError {
	p, err := parseURL(url)
	if err != nil {
		return err
	}
	if p.scheme == "http" {
		return ssrfError("http scheme blocked")
	}
	if !hostAllowed(p.host, p.path) {
		return ssrfError(fmt.Sprintf("host %s not on allowlist", p.host))
	}
	// Reject literal IP hosts.
	if ip := net.ParseIP(p.host); ip != nil && ipIsPrivate(ip) {
		return ssrfError("literal IP host")
	}
	return nil
}

// ── Template expansion ────────────────────────────────────────────────────────

// ExpandTemplate replaces known placeholders with context values.
// Mirrors webhook.rs::expand_template.
func ExpandTemplate(template string, ctx *WebhookContext) string {
	r := strings.NewReplacer(
		"{session_id}", ctx.SessionID,
		"{tool}", ctx.Tool,
		"{cost}", fmt.Sprintf("%.4f", ctx.Cost),
		"{summary}", ctx.Summary,
	)
	return r.Replace(template)
}

// ── Retry loop ────────────────────────────────────────────────────────────────

const maxAttempts = 3

var retryDelaysMS = []int64{1000, 2000, 4000}

// DispatchWithRetry runs the transport with backoff.
// sleeper is injectable for tests. Mirrors webhook.rs::dispatch_with_retry.
func DispatchWithRetry(
	transport WebhookTransport,
	url, body string,
	sleeper func(time.Duration),
	stats *RetryStats,
) *WebhookError {
	for attempt := 0; attempt < maxAttempts; attempt++ {
		stats.Attempts.Add(1)
		switch transport.Send(url, body) {
		case AttemptSuccess:
			return nil
		case AttemptPermanent:
			return statusError(0)
		case AttemptRetryable:
			if attempt+1 >= maxAttempts {
				return networkError("Exceeded max retry attempts")
			}
			delay := retryDelaysMS[attempt]
			sleeper(time.Duration(delay) * time.Millisecond)
		}
	}
	return networkError("retry exhausted")
}

// DispatchWebhook validates, expands the template, then dispatches.
// Mirrors webhook.rs::dispatch_webhook.
func DispatchWebhook(transport WebhookTransport, endpoint *WebhookEndpoint, ctx *WebhookContext) *WebhookError {
	if err := CheckSSRF(endpoint.URL); err != nil {
		return err
	}
	body := ExpandTemplate(endpoint.Template, ctx)
	stats := &RetryStats{}
	return DispatchWithRetry(transport, endpoint.URL, body, func(d time.Duration) { time.Sleep(d) }, stats)
}

// ── Default (real) HTTP transport ────────────────────────────────────────────

// HTTPTransport sends a real POST using stdlib net/http.
// This replaces Rust's LoggingTransport stub: in the Go port we have
// a concrete net/http client available.
type HTTPTransport struct {
	client *http.Client
}

// NewHTTPTransport creates a transport with a 10-second timeout. Redirects are
// NOT followed: CheckSSRF validates only the initial URL, so following a 3xx to
// an internal address would be a redirect-SSRF hole. A redirect surfaces as its
// 3xx response (→ AttemptPermanent).
func NewHTTPTransport() *HTTPTransport {
	return &HTTPTransport{
		client: &http.Client{
			Timeout: 10 * time.Second,
			CheckRedirect: func(_ *http.Request, _ []*http.Request) error {
				return http.ErrUseLastResponse
			},
		},
	}
}

func (t *HTTPTransport) Send(url, body string) AttemptOutcome {
	req, err := http.NewRequest(http.MethodPost, url, bytes.NewBufferString(body))
	if err != nil {
		return AttemptPermanent
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("User-Agent", "claude-devtools/webhook")

	resp, err := t.client.Do(req)
	if err != nil {
		return AttemptRetryable
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 200 && resp.StatusCode < 300 {
		return AttemptSuccess
	}
	if resp.StatusCode >= 500 {
		return AttemptRetryable
	}
	return AttemptPermanent
}
